server/app/auth/oidc/CiviformOidcProfileCreator.java

```diff
@@ -138,9 +148,35 @@ protected CiviFormProfileData mergeCiviFormProfile(
 
     civiformProfile.getProfileData().addAttribute(CommonProfileDefinition.EMAIL, emailAddress);
 
+    Optional<String> sessionId = getSessionId(context);
+    if (sessionId.isPresent() && enhancedLogoutEnabled(context)) {
+      // Save the id_token from the returned OidcProfile in the account so that it can be
+      // retrieved at logout time.
+      civiformProfile
+          .getAccount()
+          .thenAccept(
+              account -> {
+                SerializedIdTokens serializedIdTokens = account.getSerializedIdTokens();
+                if (serializedIdTokens == null) {
+                  serializedIdTokens = new SerializedIdTokens();
+                  account.setSerializedIdTokens(serializedIdTokens);
+                }
+                IdTokens idTokens = idTokensFactory.create(serializedIdTokens);
+                idTokens.purgeExpiredIdTokens();
+                idTokens.storeIdToken(sessionId.get(), oidcProfile.getIdTokenString());
+                account.save();
+              })
+          .join();
+    }
+
     return civiformProfile.getProfileData();
   }
 
+  private Optional<String> getSessionId(WebContext context) {
+    PlayWebContext playWebContext = (PlayWebContext) context;
+    return playWebContext.getNativeSession().get(SessionIdFilter.SESSION_ID);
```

Comment:
The two usages are slightly different (this one uses the `identityProviderType()` method rather than a passed-in parameter). The static method would need to have the `SettingsManifest` passed in as well. I'm not sure it's a win.